metamodel/processor/src/main/java/org/hibernate/search/metamodel/processor/impl/IndexedEntityMetamodelAnnotationProcessor.java

```diff
@@ -0,0 +1,150 @@
+	private final HibernateSearchMetamodelProcessorContext context;
+	private final ProcessorIntrospectorContext introspectorContext;
+	private ProcessorPojoModelsBootstrapIntrospector introspector;
+
+	public IndexedEntityMetamodelAnnotationProcessor(HibernateSearchMetamodelProcessorContext context) {
+		this.context = context;
+		this.introspectorContext = new ProcessorIntrospectorContext( context, new HashMap<>() );
+	}
+
+	@Override
+	public void process(RoundEnvironment roundEnv) {
+		TypeElement indexedAnnotation = context.elementUtils().getTypeElement( ANNOTATION_INDEXED );
+		Set<? extends Element> indexedEntities = roundEnv.getElementsAnnotatedWith( indexedAnnotation );
+
+		try ( CloseableSearchMapping searchMapping = StandalonePojoIntegrationBooter.builder()
+				.introspectorCustomizer( this::wrapIntrospector )
+				.property( "hibernate.search.backend.directory.type", "local-heap" )
+				.property(
+						StandalonePojoMapperSettings.MAPPING_CONFIGURER,
+						BeanReference.ofInstance( (StandalonePojoMappingConfigurer) configurationContext -> {
+							// disable any discovery by annotations -- that won't work as we do not implement annotations in the pojo model
+							configurationContext.annotationMapping()
+									.discoverAnnotatedTypesFromRootMappingAnnotations( false )
+									.discoverAnnotationsFromReferencedTypes( false )
+									.discoverJandexIndexesFromAddedTypes( false )
+									.buildMissingDiscoveredJandexIndexes( false );
+
+							ProgrammaticMappingConfigurationContext programmaticMapping =
+									configurationContext.programmaticMapping();
+
+							ProcessorAnnotationProcessorContext ctx = new ProcessorAnnotationProcessorContext(
+									context.elementUtils(),
+									context.typeUtils(), context.messager(), programmaticMapping
+							);
+
+							for ( Element el : indexedEntities ) {
+								if ( el.getKind().isClass() || el.getKind().isInterface() ) {
+									TypeElement indexedEntityType = (TypeElement) el;
+
+									String typeName = indexedEntityType.getQualifiedName().toString();
+
+									introspectorContext.typeElementsByName().clear();
+									introspectorContext.typeElementsByName().put( typeName, indexedEntityType );
+
+									TypeMappingStep typeMappingContext = programmaticMapping.type( typeName );
+									typeMappingContext.indexed().enabled( true );
+									typeMappingContext.searchEntity().name( typeName );
+
+									processTypeAndProperties( indexedEntityType, typeMappingContext, ctx );
+								}
+								else {
+									// TODO: generate message bundle with JBoss Logging ?
+									context.messager()
+											.printMessage(
+													Diagnostic.Kind.ERROR,
+													"Unexpected location of the " + indexedAnnotation.getQualifiedName()
+															+ ". Expected to be placed on an indexed type.",
+													el
+											);
+								}
+							}
+
+						} )
+				).build()
+				.boot() ) {
```

Comment:
The main idea behind this approach is that we are starting a standalone mapper with all possible "annotation discovery" disabled since an annotation processor won't play nice with here 😃.
Then there is a mapping configurer that iterates through the `@Indexed ` entities and uses programmatic mapping to build the model.